### PR TITLE
daemon: Allow Proxy Configuration in daemon.json

### LINKED ISSUE #26234: Introducing global environment variables to the docker-engine (flag,docker.config)
**Problem Statement**

I would like to make sure that a restarting zookeeper service reuses a certain ID after the task is rescheduled. For this the task has to hold on a piece of information that determines his current host.

**Current Workaround**
I achieve this currently by passing in `/etc/hostname:/etc/docker-hostname:ro`, which allows me to grab the hostname of the node the container is running on and use this as a key in a key/value store which holds all Zookeeper-IDs.

The script to create a unique ID (https://github.com/qnib/consul-content/blob/master/consul/bin/get-uniq-id.sh#L6-L7), will use `/etc/hostname` as a key, if it is present.

```
elif [ -f /etc/docker-hostname ];then
HOSTNAME=$(tail -n1 /etc/docker-hostname)
```
I then start a new service like this:
```
docker service create --name zookeeper -e DC_NAME=dc1 -e CONSUL_CLUSTER_IPS=consul -e CONSUL_SKIP_CURL=true -e ZK_USE_CONSUL=true \
                      -e CONSUL_GETHOSTNAME=true --mode global  --network consul-net --publish 2181:2181 \
                      --mount type=bind,source=/etc/hostname,target=/etc/docker-hostname:ro \
                      --mount type=bind,source=/tmp/,target=/data/ \
                      qnib/alpn-zookeeper
```

This breaks on hosts not providing `/etc/hostname`, for which I have to determine before hand, where they store the hostname.
And it is an semi-ugly workaround anyway...

**Proposal**
I would like to be able to define environment variables in the `dockerd` startup, which are passed to each container started by that given docker-engine.

```
dockerd --global-env=DOCKER_HOSTNAME:physical1 --global-env=RACK:corner2
```

By doing so I do not have any constraints in my `docker run`, `docker-compose up`, `swarmctl creat` or `docker service create` command, as this will be magically introduced by the docker-engine on each host of my swarm cluster.

### PR DESCRIPTION
**Pull request by dave-tucker**
_Updated 30/01/2016 to reflect the current state of this PR_

**\- What I did**

Added a new configuration option to allow proxy settings to be configured at the daemon level and inherited by every container that's run.

**\- How I did it**

Added a new `proxyEnv` field to the `Daemon` struct and piggybacked on the existing implementation of environment injection used by `links`.

**\- How to verify it**

There are tests! But you can manually verify by:
1) Setting `proxy-env` in your `daemon.json`
2) `docker run --rm -it alpine env`

**\- Description for the changelog**

Allowed a set of environment variables to be configured that allow the setting of HTTP/HTTPS/FTP proxy settings of every container run on this host.

**\- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by duglin** [comment]
What's the usecase for this?

**Comment by dave-tucker** [comment]
@duglin it's described quite well in #26234
but you may, for example, wish to let your containers know a little about where they are running.
a concrete example would be that some big data applications like to know which rack, aisle etc.. that they are being run on...
you could also use this to propagate http proxy settings in to your containers if you are operating in an environment where you *know* upfront that any containers run will require a proxy to be able to access WWW.

**Comment by dave-tucker** [comment]
A note to maintainers... I'm still working on this, but pushing very early to try and get the tests running  on CI as building locally is proving difficult for some reason

**Comment by AkihiroSuda** [comment]
Can we use daemon labels instead of daemon env vars?
 (It will be exposed to containers via on-going introspection filesystem https://github.com/docker/docker/pull/24893 )

For proxy settings, I think perhaps it should be tied to a network rather than the daemon

**Comment by duglin** [comment]
Thinking about this... I'm wondering about people who are very sensitive to the env vars in their containers - they might be annoyed with new ones showing up that they have no control over. One of the benefits of containers/images is that I can know for sure exactly what the container will look like.  With this I lose that control. I can see providers injecting all sorts of things into people containers with this feature and container owner can't stop it.

Today new env vars can be injected (via `-e` on `docker run`) but that's a opt-in approach - which is good.  I wonder if this should be an opt-in thing as well via something like a `--system-envs` flag.  I thought about a flag to disable it, but the default should be to NOT muck with people.

**Comment by justincormack** [comment]
@duglin if you set things on the daemon, that is opt in too as far as I see it. Ok it is one step removed, but we allow a lot of defaults to be set there. I think the idea here was you would be able to unset them with `-e var=` at run time.

**Comment by duglin** [comment]
That's a different level of opt-in sure, but then you would require each person (on each host) to modify their run cmd based on that host's daemon env vars.  Keep in mind there are envs where the person doing the 'docker run' isn't the same person who owns the machine so they have no idea what daemon-level env vars might be set until its too late. And in a swarm setup if there are variants between each host's env var list (which I hope wouldn't be true, but ya never know) then since the host my app is put is on is random, there may be no way for me to unset things in a consistent way.

Client-side "docker run" and config file options to specify this (either opt-in or out depending on which default is chosen at the host level) would really help and give the person managing the **container** the control they need.

**Comment by thaJeztah** [comment]
We discussed this in a maintainers review session, and we're not sure about this; if the introspection API is implemented, this would probably allow the daemon to obtain this information as well, and may be a cleaner approach (https://github.com/docker/docker/issues/26331). Would that solve the use case?

Also, perhaps this is something for Swarm (dynamically set environment variables, based on where a container is scheduled?)

**Comment by thaJeztah** [comment]
ping @dave-tucker ^^ wdyt?

**Comment by dave-tucker** [comment]
@thaJeztah Introspection doesn't work for this use case... Unfortunately, only environment variables will cut it as this is what the applications are already expecting.

While it could theoretically be implemented in Swarm it wouldn't fix the use case for a single daemon... unless if were a Swarm of one of course.

**Comment by dave-tucker** [comment]
Update this PR. Having some trouble with `integration-cli`.
While it looks like the daemon is started with the correct arguments, the environment variables aren't appearing in containers. When I run this manually it works 100% of the time.
Could be an environment issue on my end so will let CI have the final say ;)

**Comment by LK4D4** [comment]
@thaJeztah time to discuss again?
I'm quite reluctant to this because I think new features is extremely heavy on us and use-case in #26234 does not require daemon-wide option for sure.

**Comment by thaJeztah** [comment]
@LK4D4 @dave-tucker yeah, I'm not convinced on the "any env var" use case. We were discussing a use case for allowing just `HTTP_PROXY` , `HTTPS_PROXY` etc, which may make more sense (my 0.02c).

For example, what happens if I set `MYSQL_ROOT_PASSWORD` as a daemon level env-var, and I run the official MySQL image? https://hub.docker.com/_/mysql/

**Comment by thaJeztah** [comment]
We discussed this in the maintainers meeting, and we are ok with adding just the `HTTP_PROXY` (etc.)  on the daemon; basically we want to limit to the default options that `--build-arg` supports.

We do have one concern; how can we unset an environment variable on a single container? We need to have a solution for that before merging / going ahead

**Comment by thaJeztah** [comment]
@tianon just came up with another question; are these env-vars baked in into images I build on that daemon? For build, we have `--build-args` for that; how does this PR handle this?

**Comment by dave-tucker** [comment]
@thaJeztah simple, order of overrides is as follows:

`daemon global -> image -> run`

So to override you may either bake an env variable in to an image or supply it at runtime with `-e`

I hadn't considered the `build-args` interaction.
In that case I would suggest that it were to support a similar override behavior:

`daemon global -> build-args`

IMO, the resulting image should not bake in the `daemon global` variables. Does this happen today with `build-args`?

**Comment by thaJeztah** [comment]
> simple, order of overrides is as follows:

We should look at _unsetting_ though; `-e FOO=""` is not the same as `FOO` not being set. (And, no `-e FOO` doesn't work, because that copies the value of `$FOO` in the environment if it's exported/set)

> IMO, the resulting image should not bake in the daemon global variables. Does this happen today with build-args?

No, build-args are specifically designed to not be persisted in the image; that way you can set a proxy during build, but don't have it persisted in the final image. build-args _are_ visible in the image's history though (something to take into account), because they are prefixed for each run;

```
docker history --no-trunc foo

IMAGE                                                                     CREATED             CREATED BY                                                                                          SIZE                COMMENT
sha256:317b6d045ac269d65f9915a02cbef61802957ab79a96012325926dd45c87197d   17 seconds ago      |1 HTTP_PROXY=user:pass@myproxy.com /bin/sh -c echo "hello"                                         0 B
sha256:e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba   5 weeks ago         /bin/sh -c #(nop)  CMD ["sh"]                                                                       0 B
<missing>                                                                 5 weeks ago         /bin/sh -c #(nop) ADD file:ced3aa7577c8f970403004e45dd91e9240b1e3ee8bd109178822310bb5c4a4f7 in /    1.093 MB
```

They don't persist in the image layers, unless I make them (e.g. `ENV FOO=$HTTP_PROXY`)

**Comment by dave-tucker** [comment]
@thaJeztah thanks.
You are right re: *unsetting*, I'd forgotten that `-e FOO=""` is not equivalent.
Perhaps `-e !FOO` might be a good way of un-setting it. WDYT?

In relation to `--build-args` I'm not sure how to proceed.

We can either:
a) Disable this feature for build.
b) Allow this feature to replace `build-args` for proxy variables *if they are configured on the daemon*. This is backwards compatible as `build-args` will override the daemon arguments anyway, but IMO would be a major UX benefit - especially if you don't know that your daemon is behind a proxy.

My original intention was a), but I'd be happy to look at implementing b).

**Comment by vdemeester** [comment]
ping @dave-tucker @thaJeztah @tianon what is the status here ?

**Comment by dave-tucker** [comment]
@vdemeester well, the code has bit-rotted quite substantially since I opened the PR 😨
I'm going to try to update this next week and incorporate all feedback so far so we can look at the design again...

**Comment by alexellis** [comment]
Reference - Improve UX around http-proxy for those in corporate environment #30323

**Comment by LK4D4** [comment]
@thaJeztah ok, I propose to consider this last time on maintainers meeting.

**Comment by alexellis** [comment]
@LK4D4 what do you mean? That it's the last opportunity to talk about it?

**Comment by LK4D4** [comment]
@alexellis yeah, usually it should be closed after the first discussion.
It looks like we can't make progress here because maintainers doesn't like the design. Maybe they can be convinced on the meeting.

**Comment by dave-tucker** [comment]
Updated PR coming momentarily - awaiting a clean `make test` locally before pushing

**Comment by thaJeztah** [comment]
@dave-tucker CI is failing - golint issue;

```
02:41:02 Errors from golint:
02:41:02 opts/opts.go:305:1: exported function ValidateProxyEnv should have comment or be unexported
02:41:02 
02:41:02 Please fix the above errors. You can test via "golint" and commit the result.

**Comment by LK4D4** [comment]
@thaJeztah it's still in design review. Not much sense in fixing golint IMO :)

**Comment by dave-tucker** [comment]
I fixed the golint issue. The failing test is one that I wrote. 
For some reason it the test suite doesn't seem to honour the new flags that were added
If I build  a binary and test it myself, it works fine.

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -4415,3 +4415,17 @@ func (s *DockerSuite) TestRunHostnameInHostMode(c *check.C) {
 	out, _ := dockerCmd(c, "run", "--net=host", "--hostname=foobar", "busybox", "sh", "-c", `echo $HOSTNAME && hostname`)
 	c.Assert(strings.TrimSpace(out), checker.Equals, expectedOutput)
 }
+
+func (s *DockerDaemonSuite) TestRunWithProxyEnv(c *check.C) {
+	expectedOutput := "HTTP_PROXY=127.0.0.1:3001"
+	s.d.StartWithBusybox(c, "--debug", "--proxy-env", expectedOutput)
+	out, _ := dockerCmd(c, "run", "busybox", "env")
```
Comment (thaJeztah) [inline integration-cli/docker_cli_run_test.go]: should this be

```go
out, _ := s.d.Cmd("run", "busybox", "env")
```

So that it actually runs on the daemon you just started?

File: opts/opts.go [hunk 2]
```diff
@@ -302,6 +302,26 @@ func ValidateSysctl(val string) (string, error) {
 	return "", fmt.Errorf("sysctl '%s' is not whitelisted", val)
 }
 
+// ValidateProxyEnv checks proxy environment variables for validity
+func ValidateProxyEnv(val string) (string, error) {
+	permitted := []string{"HTTP_PROXY", "HTTPS_PROXY", "NO_PROXY", "FTP_PROXY"}
```
Comment (AkihiroSuda) [inline opts/opts.go]: `ALL_PROXY`, `RSYNC_PROXY`, .. (AFAIK)

File: opts/opts.go [hunk 3]
```diff
@@ -302,6 +302,26 @@ func ValidateSysctl(val string) (string, error) {
...
+	matched := false
+	for _, p := range permitted {
+		if strings.HasPrefix(val, p) {
```
Comment (AkihiroSuda) [inline opts/opts.go]: Why prefix?

File: opts/opts.go [hunk 4]
```diff
@@ -302,6 +302,26 @@ func ValidateSysctl(val string) (string, error) {
...
+			matched = true
+			break
+		}
+		if strings.HasPrefix(val, strings.ToLower(p)) {
+			matched = true
+			break
+		}
+	}
+	if matched == false {
```
Comment (AkihiroSuda) [inline opts/opts.go]: `matched == false` --> `!matched`

File: integration-cli/daemon/daemon.go [hunk 5]
```diff
@@ -250,6 +250,7 @@ func (d *Daemon) StartWithLogFile(out *os.File, providedArgs ...string) error {
 	}
 
 	args = append(args, providedArgs...)
+	fmt.Println(args)
```
Comment (AkihiroSuda) [inline integration-cli/daemon/daemon.go]: Please remove

File: cli/command/system/info.go [hunk 6]
```diff
@@ -319,6 +319,13 @@ func prettyPrintInfo(dockerCli *command.DockerCli, info types.Info) error {
 
 	fmt.Fprintf(dockerCli.Out(), "Live Restore Enabled: %v\n", info.LiveRestoreEnabled)
 
+	if info.ProxyEnv != nil && len(info.ProxyEnv) > 0 {
```
Comment (AkihiroSuda) [inline cli/command/system/info.go]: Since `len(nil)` always return 0, you don't need to check nil here

File: opts/opts.go (hunk 2 above)
Comment (thaJeztah) [inline opts/opts.go]: I think this is the current list we support for `--build-arg`; if we decide to add more, we should keep those in sync (should be a separate PR / discussion) ; maybe we can use the same list (from code) to guarantee they stay in sync?

Note that these are (afaik) not standardized and there's variations of these (uppercase/lowercase); some software uses lowercase, other uppercase 😞

Comment (thaJeztah) [inline opts/opts.go]: List for build-args is here; https://github.com/docker/docker/blob/597843e79040a4f00bfe2abd83343d43edd195f5/builder/dockerfile/builder.go#L40-L50

Comment (AkihiroSuda) [inline opts/opts.go]: OK

File: cli/command/system/info.go (hunk 6 above)
Comment (dave-tucker) [inline cli/command/system/info.go]: Yes I do, because `len([]string{}) == 0` and we don't want to print the information if there is none to show.

File: integration-cli/docker_cli_run_test.go (hunk 1 above)
Comment (dave-tucker) [inline integration-cli/docker_cli_run_test.go]: 😚 that. is. amazing. it's bothered me since the dawn of this PR and I hadn't even noticed this hiding in plain sight. THANK YOU!

**Comment by dave-tucker** [comment]
Addressed feedback from @AkihiroSuda and @thaJeztah.
My tests are failing now with `docker run` exiting 125... still debugging my end but hopeful that a solution is in sight

File: cli/command/system/info.go (hunk 6 above)
Comment (dave-tucker) [inline cli/command/system/info.go]: deleted my comment as I see what you mean now :)

File: integration-cli/docker_cli_run_test.go (hunk 1 above)
Comment (thaJeztah) [inline integration-cli/docker_cli_run_test.go]: The amount of docker-in-docker-and-other-daemons in CI is quite confusing :smile:

Comment (vdemeester) [inline integration-cli/docker_cli_run_test.go]: Yeah, it bothers me that we have two ways to write test depending on if we start a daemon or use the "current" daemon. I'm working on fixing that (so it's simpler to write tests)

**Comment by dave-tucker** [comment]
Running tests locally...

```
PASS: docker_cli_run_test.go:4419: DockerDaemonSuite.TestRunWithProxyEnv	1.835s
PASS: docker_cli_run_test.go:4427: DockerDaemonSuite.TestRunWithProxyEnvOverride	1.855s
OK: 2 passed
PASS
``

hmmmm.....

**Comment by alexellis** [comment]
Remind me of the Docker make command to run the tests for a specific package?

**Comment by vdemeester** [comment]
for integration tests : `TESTFLAGS="-check.f DockerDaemonSuite.TestRun" make test-integration-cli` will run all tests in the `DockerDaemonSuite` starting by `TestRun`.
@alexellis ^^

**Comment by dave-tucker** [comment]
Excuse my clumsy rebasing

**Comment by thaJeztah** [comment]
@vdemeester the `TESTFLAGS=..` should go _after_ `make` (before usually works, but I think I've seen cases where it made a difference)

**Comment by dave-tucker** [comment]
Windows CI failures appear unrelated

**Comment by thaJeztah** [comment]
Left a comment in https://github.com/docker/docker/pull/30588#issuecomment-283106005

**Comment by thaJeztah** [comment]
Closing this one as well for now (https://github.com/docker/docker/pull/30588#issuecomment-292198998), but we'll definitely want some options for this, so ping me as well to re-open 🤗

**Comment by thaJeztah** [comment]
I'm sure @alexellis will remind us as well :smile:
